### ISSUE TITLE
Adding support for Ali Ashraf's vector tiles

### ISSUE THREAD

**Issue by pauldendulk** [pauldendulk/Mapsui#378]
@AliFlux did some impressive work! There is a lot to discuss around this issue. I started research on this a year ago but never had the time to finish it. @charlenni Has also been working on this. Perhaps we can have a hangout session to discuss this? I am in the central european time zone and would have time Tuesday and Wednesday evening and Friday morning (and part of the afternoon).

Some reactions. 
- In the past I have used many renderers for Mapsui. I have invested into OpenTK and had SharpDX in mind for UWP. I released app with OpenTK, so it was at a fair level, but disbanded ith because the support for OpenTK was not good enough and it was hard to implement the simplest things.
- SkiaSharp is a library that is maintained by Microsoft itself. I don't know what the status of SharpDX is right now. But I think the maintenance of SkiaSharp is at a higher level.
- Skia is developed by Google. It is an important core component of Chrome and Android. Apparent limitations might be due to lack of knowledge.
- Skia does have hardware acceleration. It is not very hard to turn it on. You have to use the GL canvasses for this. It is implemented for iOS. 
- Hardware acceleration was added in a later stage to Skia so SharpDX might have better support for this. We will have to measure if the difference is significant.
- By the look of the demo (did not look at the code yet) you library renders the vectortiles to bitmap, like I did in my original sample. This means:
  - You will still see blocky graphics at some levels (-)
  - You can change the style on the fly (+)
  - You do have the advantage of smaller tiles (+)
- This functionality (mention in the previous point) is a significant addition to Mapsui and has good decoupling from Mapsui. This is a good thing.
- What @charlenni and I have been working on it direct rendering of vector on the Mapsui canvas. 
- Mapsui has been designed to have different rendering implementations. Supporting the full functionality of a renderer is a lot of work. I don't think that adding a SharpDX renderer is worth it. 
- My vectortile implementation is in the vectortiles branch of Mapsui. There is also a 'performance' branch where I added tweaks to improve performance (but not yet hardware acceleration).
- Mapsui is designed to fetch tiles in an offline thread, this is why the ui is still smooth while parsing and rendering the tiles takes a lot of time. Using more threads during this tile building will improve performance. It is a performance optimization that will not effect Mapsuis architecture.
- Mapsui uses BruTile for tile support. It supports any kind of tile schema. 512x515 tiles are just a different schema definition, it will work in Mapsui if you define it correctly in the BruTile tile schema.

**Comment by charlenni** [comment]
* SharpDX uses DirectX in background, so it is limited to Windows platforms. One big advantage of Mapsui is, that it runs on many platforms (Windows, Linux, Mac OS, Android, iOS). With this in mind, OpenGL could be the only additional renderer for Mapsui, because it is supported on all platforms. One possible library would be [OpenGL.Net](https://github.com/luca-piccioni/OpenGL.Net). Works on all platforms except iOS.
* The problem with Skias hardware acceleration is, that it doesn't support paths. If I'm right, than paths are rendered on CPU to bitmaps  and than uploaded to GPU. And other drawing methods like DrawLine, which are completely  executed on GPU don't support things like stroke join :(. So Skia hardware acceleration isn't the gral.

This said, there could be only one choice for an additional renderer: OpenGL .

**Comment by pauldendulk** [comment]
Thanks for that info charlenni. OpenGL.net supports UWP as well? I guess so since it supports net core. Is iOS supoort expected at some point?

For me SharpDX was relevant because UWP had no option for OpenGL back then.

With OpenTK line joins were not supported at all. You had to fix that with your own microscopic polygon rendering. You had to invent you own algorithms to draw just a polygon with holes. OpenGL net won t be much easier to use. If you get this stuff implemented it is still a question if it will be faster than skia.

Then again. Always good to research this. I should the renderers injectable so external renderers can be used.

**Comment by xivk** [comment]
Have you seen this:

https://github.com/mellinoe/veldrid

It may solve some of the issues.

**Comment by charlenni** [comment]
Looks promising, but there is still the problem, who of use has so much knowledge, that he/she could write a renderer for one of the systems?

**Comment by AliFlux** [comment]
@xivk, I didn't check this library but it looks pretty low level by the docs. Will check it soon. What we need is a high level hardware accelerated 2D drawing library. Something like [Win2D](https://github.com/Microsoft/Win2D), but for all platforms.

There's also some discussion on integrating [SkiaSharp](https://github.com/mono/SkiaSharp/issues/243) with [Angle](https://github.com/google/angle), which is Google's backend for cross platform graphics. Another thing that can be done is to bundle OpenGL context with the existing skiasharp library. [PixelFarm](https://github.com/PaintLab/PixelFarm) has been doing it.

**Comment by charlenni** [comment]
Perhaps [this](https://softwareengineering.stackexchange.com/questions/191472/why-have-hardware-accelerated-vector-graphics-not-taken-off) articel brings some light in the discussion, why not allways a GPU solution is the best.

**Comment by pauldendulk** [comment]
The rendering in @AliFlux library happens in a background thread (when called from Mapsui). Performance optimization is not important here. The first thing to do would be to add hardware acceleration to skiasharp. The limitations of skiasharp are theoretical so far. I have not read anything indicating it is slower than other hardware accelerated solution. I have used many rendering engines and they all come with their own limitations. SkiaSharp is a library with good support and documentation. It is the most pleasant library I worked with so far.

**Comment by AliFlux** [comment]
I started the VTR library with OpenTK canvas backend. But the polygon result was horribly aliased, so I switched to SkiaSharp.

**Comment by memsom** [comment]
So - I looked at @AliFlux's work and it is indeed impressive. I have a requirement to make this work on mobile, so I have removed the parts that depend on the System.Windows.Media / Imaging namespaces and replaced them with pure SkiaSharp, and it does seem to work so far. I feel like it is marginally slower under Windows. But as this is aimed at Android I guess with the caching is should be okay. Has anyone else looked at removing the Windows dependencies?

So, what I did was - in the ICanvas, we now have
```csharp
        byte[] FinishDrawing();
```
Then rather than `BitmapSources` we just render the canvas to a PNG directly and grab the raw bytes:
```csharp
public byte[] FinishDrawing()
{
    using (var image = surface.Snapshot())
    using (var data = image.Encode(SKEncodedImageFormat.Png, 80))
    using (var result = new MemoryStream())
    {
        data.SaveTo(result);
        return result.ToArray();
    }
}
```
Then everything to do with the caching and passing back of data just uses those bytes.
I don't know if it can be optimised more, but something is better than nothing. Doing this small change will hopefully make it cross platform and will work for my project.

**Comment by Feroks** [comment]
@memsom I think the biggest problem was rendering street names. Especially when it needs to be curved or when it is located in between tiles. If i recall correctly @AliFlux did not implement it.

**Comment by charlenni** [comment]
Yes, there are two problems: First problem is, that the label handling is more than basic. Only a few point labels are drawn. The second problem is, that it is only a tile renderer. What you get at the end is a raster tile with all the disadvantages of raster tiles.

What I don’t remember is, if symbols are drawn.

I made many PRs to push Mapsui in the right direction to use it for real vector maps, but struggled always at the label problem. It seems beyond my capabilities.

**Comment by memsom** [comment]
For my use - the tiles are not an issue. We need offline maps and vector is the format we are told is best. I already invested quite heavily in to customising Mapsui for our needs, so this works quite nicely. Also a reason to update our internal version inline. (Most of our changes were to removing things, adding in the ability to have more control over layers and the Z order and making private things protected so that we can access them in our internal library that sits on top of Mapsui. The majority of changes are published on my github account, the ones not yet published have not been released.)

So - I have managed to abstract out all of the Windows specific stuff and now have a version that will run with the 3.02 release of Mapsui. Hopefully at some point before the end of next week, I will have it working under Xamarin Forms at least for Android. The main challenges now are moving the styles and fonts to be resources I think. I also want to make sure the PBF stuff still works as taking raw OpenStreetmap data is pretty powerful IMO.

The master on my branch has most of the generic changes in it, the one called XamarinForms is at the stage where I could start adding in the control and the glue code. I will solve the resources first though as messing with Windows test apps is less painful than Andoid.
